### PR TITLE
feat(nuxt): enable preloading object-syntax routes

### PR DESCRIPTION
**Pull request by Mini-ghost**
### 🔗 Linked issue

- #18845 

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

After this PR, the path parsed by the following code all will be `/home`. As a result, the preload function will not retry and will instead return early.

```html
<!-- Passing a string -->
<NuxtLink to="/home"> Home </NuxtLink>

<!-- Passing a string with query string -->
<!-- `router.resolve('/home?year=2023').path` is `/home` -->
<NuxtLink to="/home?year=2023"> Home </NuxtLink>

<!-- Passing a object -->
<!-- `router.resolve({ name: 'home', query: { year: '2023' } }).path` is `/home` -->
<NuxtLink :to="{ name: 'home', query: { year: '2023' } }"> Home </NuxtLink>
```

In addition, when users use `preloadRouteComponents`, they are not limited to type string.

```ts
const route: RouteLocationRaw = {
  name: 'dashboard'
}

preloadRouteComponents(dashboard)
const submit = async () => {
  const results = await $fetch('/api/authentication')
  if (results.token) {
    await navigateTo(dashboard)
  }
}
```
### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/app/composables/preload.ts [hunk 1]
```diff
@@ -34,11 +34,13 @@ function _loadAsyncComponent (component: Component) {
...
 
-export async function preloadRouteComponents (to: string, router: Router & { _routePreloaded?: Set<string>; _preloadPromises?: Array<Promise<any>> } = useRouter()): Promise<void> {
+export async function preloadRouteComponents (to: RouteLocationRaw, router: Router & { _routePreloaded?: Set<string>; _preloadPromises?: Array<Promise<any>> } = useRouter()): Promise<void> {
   if (process.server) { return }
 
+  const { path, matched } = router.resolve(to)
+
   if (!router._routePreloaded) { router._routePreloaded = new Set() }
-  if (router._routePreloaded.has(to)) { return }
+  if (router._routePreloaded.has(path)) { return }
```
Comment (danielroe) [inline packages/nuxt/src/app/composables/preload.ts:43]: maybe we can also early-return if there are no matched components?

Comment (Mini-ghost) [reply packages/nuxt/src/app/composables/preload.ts:43]: Like this?

```ts
if (!matched.length) { return }
if (!router._routePreloaded) { router._routePreloaded = new Set() }
if (router._routePreloaded.has(path)) { return }
```

I think it looks better this way, but I'm not sure when there is no matching component?

Comment (danielroe) [reply packages/nuxt/src/app/composables/preload.ts:43]: Looks perfect!

**Comment by danielroe** [comment]
Would you update this following the merge of your other PR? 🙏

**Comment by Mini-ghost** [comment]
I resolved the conflict caused by #19144.

**Comment by Mini-ghost** [comment]
I'm unsure which option is better between using `name` or `path` for dynamic route, Even though they resolve to the same page components, their paths may still differ. While `name` may not have this issue, it is possible for it to be `undefined` or `null`.

Does the `name` of the route always exist in Nuxt? If so, would it be better to use it?

Overall, this issue is not significant because although dynamic routing repeats preloading, it doesn't require sending additional network requests.

**Comment by danielroe** [comment]
`name` does not always exist, so I think better to use `path`.

**Pull request merged**